imports/plugins/included/simple-inventory/server/no-meteor/mutations/updateSimpleInventoryBulk.js

```diff
@@ -0,0 +1,97 @@
+import _ from "lodash";
+import Logger from "@reactioncommerce/logger";
+import { inputSchema } from "../utils/defaults";
+import getModifier from "../utils/getModifier";
+
+const logCtx = { name: "simple-inventory", file: "updateSimpleInventoryBulk" };
+
+/**
+ * @summary Updates SimpleInventory data for multiple products
+ * Note/Warning: This function is only available on `context` for internal calls only - it is NOT exposed through GraphQL.
+ * As such there is no permission checks added. If this function is exposed later, it should be updated to check permissions.
+ * @param {Object} context App context
+ * @param {Object} input Input
+ * @param {Object} input.updates Array of objects, each containing fields for productConfiguration, shopId, canBackorder, inventoryInStock, isEnabled, lowInventoryWarningThreshold. See definition of these values on single updateSimpleInventory mutation.
+ * @return {Object} Empty object, or an object with a `failedOps` field containing any failed update operations
+ */
+export default async function updateSimpleInventoryBulk(context, input) {
+  const { appEvents, collections, userId } = context;
+  const { SimpleInventory } = collections;
+
+  // generate mongo operation object for each product input. Return null on any invalid input
+  let bulkOperations = input.updates.map((update) => {
+    let modifier = {};
+    try {
+      inputSchema.validate(update);
+      modifier = getModifier(update);
+    } catch (error) {
+      Logger.error({ ...logCtx, error });
+      return null;
+    }
+
+    if (!modifier.$set || !modifier.$setOnInsert) return null;
+    const { productConfiguration, shopId } = update;
+
+    return {
+      updateOne: {
+        filter: {
+          "productConfiguration.productVariantId": productConfiguration.productVariantId,
+          shopId
+        },
+        update: modifier,
+        upsert: true
+      }
+    };
+  });
+
+  bulkOperations = bulkOperations.filter((op) => op !== null);
```

Comment:
I read this to mean if an update doesn't validate, we ignore it and swallow the error. Seems like we should include a `invalidUpdates` property on the returned result summary object. Swallowing errors makes upstream problems harder to detect and debug.

Reply:
I agree, and I also think `failedOps` should have the same original `productConfiguration` that was passed in on each one. So altogether, maybe a function return object something like this:

```js
{
  failedUpdates: [
    {
      productConfiguration,
      writeError
    }
  ],
  invalidUpdates: {
    productConfiguration,
    validationError
  }
}
```